products/bridge/smart-contracts/foundry/test/Relayer/Fees.t.sol

```diff
@@ -0,0 +1,356 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity 0.8.20;
+
+import "forge-std/Test.sol";
+import {RelayerTestFixture, RelayerHarness, ValidatorManager, Test, TransferReentrancyTester} from "./Helpers.sol";
+import {IRelayerErrors} from "contracts/Relayer.sol";
+
+contract RelayerFeeTester is RelayerHarness {
+    bool public writeSuccessful;
+    address public sponsor;
+
+    constructor(
+        ValidatorManager validatorManger
+    ) RelayerHarness(validatorManager) {}
+
+    function test(address _sponsor) external meterFee(_sponsor) {
+        writeSuccessful = true;
+    }
+
+    function setSponsor(address _sponsor) external {
+        sponsor = _sponsor;
+    }
+
+    function test() external meterFee(sponsor) {
+        writeSuccessful = true;
+    }
+
+    function test(
+        address _sponsor,
+        bytes[] calldata data
+    ) external meterFee(_sponsor) {
+        data;
+        writeSuccessful = true;
+    }
+
+    function test(
+        address _sponsor,
+        uint[] calldata data
+    ) external meterFee(_sponsor) {
+        data;
+        writeSuccessful = true;
+    }
+}
+
+contract Fees is RelayerTestFixture {
+    using stdStorage for StdStorage;
+
+    RelayerFeeTester relayerTester;
+    address sponsor1 = vm.addr(1000);
+    address sender1 = vm.addr(2000);
+
+    function setUp() public {
+        relayerTester = new RelayerFeeTester(validatorManager);
+    }
+
+    function verifyFeeInvariant(
+        uint initialFeeDeposit,
+        uint gasSpent
+    ) internal {
+        // feeDeposit + feeRefund = initial deposit
+        assertEq(
+            relayerTester.feeDeposit(sponsor1),
+            initialFeeDeposit - relayerTester.feeRefund(sender1),
+            "Invariant violated: feeDeposit + feeRefund = initial deposit"
+        );
+        assertGe(
+            relayerTester.feeRefund(sender1),
+            gasSpent,
+            "Invariant violated: Sender should be refunded more than the gas spent"
+        );
+    }
+
+    function test_meterFee_happyPath() external {
+        vm.txGasPrice(10 gwei);
+        uint initialFeeDeposit = 100000 ether;
```

Comment:
You can change it to `100_000` for better readability like you did everywhere else